crates/red_knot_python_semantic/src/module_name.rs

```diff
@@ -168,6 +168,25 @@ impl ModuleName {
         };
         Some(Self(name))
     }
+
+    /// Push a new "tail" to the module name.
+    ///
+    /// The tail may be a single component, e.g. `"foo"`
+    /// or a "dotted string" consisting of multiple components,
+    /// e.g. `"foo.bar.baz"`
+    ///
+    /// ## Errors
+    /// Returns an error if any component in `tail` was not a valid identifier,
+    /// maintaining the invariant enforced by the constructor.
+    pub fn push_tail(&mut self, tail: &str) -> Result<(), &'static str> {
+        if Self::is_valid_name(tail) {
+            self.0.push('.');
+            self.0.push_str(tail);
+            Ok(())
+        } else {
+            Err("Invalid component pushed: not an identifier")
+        }
+    }
```

Comment:
We could avoid the need to return an `Error` if we make the `Argument` a `ModuleName`. 

Nit: I think `push` is sufficient because we don't allow pushing at the front and it better aligns with the `str` API. 
```suggestion
    pub fn join(&mut self, other &ModuleName {
    	self.0.push(&other.0);
    }
```

Reply:
> We could avoid the need to return an `Error` if we make the `Argument` a `ModuleName`.

Great idea!

> Nit: I think `push` is sufficient because we don't allow pushing at the front

Hmm, I don't think my current name is great but I don't like `push`, because I'd expect a `push` method to only accept a single component (like `Vec::push`), which isn't what this method does.

> it better aligns with the `str` API.

I don't think this is a great reason because the fact that this type wraps a string currently is an implementation detail. Conceptually, an absolute module name is a list of identifier strings separated by dots, so in terms of the semantics of the type it could just as easily wrap a `Vec` (or `SmallVec`, etc)

Reply:
How about `extend` or `join`. Because your concerns about `push` also apply to `push_tail`